store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/impl/FederatedGetTraitsHandler.java

```diff
@@ -16,19 +16,62 @@
 
 package uk.gov.gchq.gaffer.federatedstore.operation.handler.impl;
 
+import uk.gov.gchq.gaffer.commonutil.iterable.CloseableIterable;
+import uk.gov.gchq.gaffer.commonutil.stream.Streams;
 import uk.gov.gchq.gaffer.federatedstore.FederatedStore;
+import uk.gov.gchq.gaffer.federatedstore.operation.FederatedOperationChain;
+import uk.gov.gchq.gaffer.operation.OperationChain;
 import uk.gov.gchq.gaffer.operation.OperationException;
 import uk.gov.gchq.gaffer.store.Context;
 import uk.gov.gchq.gaffer.store.Store;
 import uk.gov.gchq.gaffer.store.StoreTrait;
 import uk.gov.gchq.gaffer.store.operation.GetTraits;
 import uk.gov.gchq.gaffer.store.operation.handler.OutputOperationHandler;
 
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
 import java.util.Set;
+import java.util.stream.Collectors;
 
+import static java.util.Objects.isNull;
+import static java.util.Objects.nonNull;
+import static uk.gov.gchq.gaffer.federatedstore.FederatedStoreConstants.KEY_OPERATION_OPTIONS_GRAPH_IDS;
+
+/**
+ * returns a set of {@link StoreTrait} that are common for all visible graphs.
+ * traits1 = [a,b,c]
+ * traits2 = [b,c]
+ * traits3 = [a,b]
+ * return [b]
+ */
 public class FederatedGetTraitsHandler implements OutputOperationHandler<GetTraits, Set<StoreTrait>> {
     @Override
     public Set<StoreTrait> doOperation(final GetTraits operation, final Context context, final Store store) throws OperationException {
-        return ((FederatedStore) store).getTraits(operation, context);
+        try {
+            FederatedOperationChain<Void, StoreTrait> wrappedFedChain = new FederatedOperationChain.Builder<Void, StoreTrait>()
+                    .operationChain(OperationChain.wrap(operation))
+                    //deep copy options
+                    .options(isNull(operation.getOptions()) ? new HashMap<>() : new HashMap<>(operation.getOptions()))
+                    .build();
+
+            final CloseableIterable<StoreTrait> concatResults = store.execute(wrappedFedChain, context);
+
+            Map<StoreTrait, Integer> rtn;
+            if (nonNull(concatResults) && nonNull(concatResults.iterator()) && concatResults.iterator().hasNext()) {
+                rtn = Streams.toStream(concatResults)
+                        //.flatMap(Collection::stream)
+                        .collect(Collectors.toMap(t -> t, ignore -> 1, (existing, replacement) -> existing + replacement));
```

Comment:
I think I understand what this is doing but it is quite hard to read as a one liner like this (at least for me). Is it turning the list of traits into a map of trait-> count?

Reply:
Yes. The 3rd parameter of toMap is responsible for resolving the merge/collision. I'm using it to maintain a count. The line below the delete everything that doesn't have count that matches the number of graphs. This is all Deprecated code and will be removed.